File: compiler/parser.py
import ast

class Parser(object):
    def __init__(self, lexer):
        self.tokens = lexer.tokenize()
        self.pos = 0
        self.root = ast.Block()

    def peek(self, i=0):
        return self.tokens[self.pos+i]

    def next(self):
        token = self.tokens[self.pos]
        self.pos += 1
        return token

    def is_next(self, name):
        return self.peek().name == name

    def match(self, name):
        token = self.next()
        if token.name != name:
            self.unexpected(name, token.name)
        return token

    def unexpected(self, expected, found):
        raise Exception("excpected %s, found %s" % (expected, found))

    def parse(self):
        self.parse_program()
        return self.root

    def parse_program(self):
        while not self.is_next("EOF"):
            expr = self.parse_expression()
            self.root.add(expr)

    def parse_block(self):
        pass

    def parse_expression(self):
        lhs = self.parse_primary()
        
        if self.is_next("OPERATOR"):
            return self.parse_expression_pre(lhs)
        else:
            return lhs

    def parse_expression_pre(self, lhs, precedence=0):
        while self.is_next("OPERATOR") and self.peek().precedence >= precedence:
            op = self.next()
            rhs = self.parse_primary()
            while self.is_next("OPERATOR") and self.peek().precedence > precedence:
                next_precedence = self.peek().precedence
                rhs = self.parse_expression_pre(rhs, next_precedence)
            lhs = ast.Operation(op.value, lhs, rhs)
        return lhs

    def parse_primary(self):
        if self.is_next("INTEGER"):
            value = self.match("INTEGER").value
            primary = ast.Integer(value)

        elif self.is_next("STRING"):
            value = self.match("STRING").value
            primary = ast.String(value)

        elif self.is_next("NAME"):
            value = self.match("NAME").value
            primary = ast.Name(value)

        else:
            primary = None
            self.unexpected("primary", self.peek().name)

        if self.is_next("LPAR"):
            args = self.parse_call_args()
            primary = ast.Message(primary, "()", args)

        return primary

    def parse_call_args(self):
        self.match("LPAR")
        args = []

        while True:
            if self.is_next("RPAR"):
                self.match("RPAR")
                return args
            else:
                args.append(self.parse_expression())

            if self.is_next("RPAR"):
                self.match("RPAR")
                return args
            elif self.is_next("COMMAN"):
                self.match("COMMA")
            else:
                self.unexpected("comma or rpar", self.peek())

File: compiler/src/parser.cc
#include "parser.hh"
#include <sstream>

Parser::Parser(Lexer* lexer) {
  this->lexer = lexer;
  lexer->tokenize();
}

Node* Parser::parse() {
  return parseProgram();
}

Node* Parser::parseProgram() {
  BlockNode* prog = new BlockNode;
  while (!lexer->isNext(Token::eof)) {
    prog->add(parseExpression());
  }
  return prog;
}

Node* Parser::parseExpression() {
  Node* lhs = parseAtom();

  if (lexer->isNext(Token::binaryop)) {
    return parseOperation(lhs, 0);
  }
  else if (lexer->isNext(Token::lpar)) {
    return parseCall(lhs);
  }
  else return lhs;
}

Node* Parser::parseOperation(Node* lhs, int precedence) {
  while (lexer->isNext(Token::binaryop)
         && lexer->peek().getPrecedence() >= precedence) {
    Token op = lexer->next();
    Node* rhs = parseAtom();
    while (lexer->isNext(Token::binaryop)
           && (lexer->peek().getPrecedence() > op.getPrecedence())) {
      int nextPrecedence = lexer->peek().getPrecedence();
      rhs = parseOperation(rhs, nextPrecedence);
    }
    lhs = new BinaryOpNode(op.getValue(), lhs, rhs);
  }
  return lhs;
}

Node* Parser::parseAtom() {
  if (lexer->isNext(Token::integer)) {
    std::string value = lexer->next().getValue();
    std::stringstream ss(value);
    int v;
    ss >> v;
    return new IntegerNode(v);
  }
  else if (lexer->isNext(Token::string)) {
    std::string value = lexer->next().getValue();
    return new StringNode(value);
  }
  else if (lexer->isNext(Token::name)) {
    std::string value = lexer->next().getValue();
    return new NameNode(value);
  }
  else if (lexer->isNext(Token::lpar)) {
    lexer->next();
    Node* expr = parseExpression();
    lexer->match(Token::rpar, ")");
    return expr;
  }
  else {
    std::cerr << "unkown token " << lexer->next().getValue() << std::endl;
    exit(0);
    return 0;
  }
}

Node* Parser::parseCall(Node* callee) {
  lexer->match(Token::lpar);

  CallNode* call = new CallNode(callee);

  while (true) {
    if (lexer->isNext(Token::rpar)) {
      break;
    }
    else {
      call->addArg(parseExpression());
      if (lexer->isNext(Token::rpar)) {
        lexer->next();
        break;
      }
      else lexer->match(Token::comma, ",");
    }
  }

  return call;
}
